src/utils/AuthProvider.tsx

```diff
@@ -0,0 +1,116 @@
+'use client';
+
+import { AuthError, AuthResponse, Session } from '@supabase/supabase-js';
+import { createContext, useContext, useEffect, useMemo, useState } from 'react';
+import { UUID } from 'crypto';
+import supabase from '@/api/supabase/createClient';
+
+// TODO: resolve inconsistent null/undefined types for empty values
+export interface AuthContextType {
+  session: Session | undefined;
+  userId: UUID | undefined;
+  userEmail: string | undefined;
+  signIn: (email: string, password: string) => Promise<AuthResponse>;
+  signOut: () => Promise<AuthError | null>;
+  signUp: (
+    email: string,
+    password: string,
+    options: object,
+  ) => Promise<AuthResponse>;
+}
```

Comment:
regarding the `TODO` - not sure how much it matters whether we're consistent with `null` or `undefined`, but I think there could be a subtle difference that can make things cleaner

for me, i see it as:
- `undefined` = the value haven't been initialized yet
- `null` = the value doesn't exist at all

ngl i think their difference is negligible and we can just use truthy value to differentiate those from usable info, but i'm curious to hear your thoughts on this

Reply:
looked more into this and there are a lot of `null` haters, specifically for javascript/typescript. Also apparently Typescrip Coding Guidelines advise against the use of `null`, and Tony Hoare—the creator of null—considers it a billion dollar mistake 💀. Here are some articles I found:
- [We don't need null ](https://lou.cx/articles/we-don-t-need-null/)
- [null vs undefined? What to choose? What to use?](https://dev.to/sduduzog/null-vs-undefined-what-to-choose-what-to-use-11g)

so maybe `undefined` is the way to go for this use case?

Reply:
I don't think we should go full on `null` or full on `undefined`. They both have their own use cases, as the second article you provided described. I would be weary of following those practices to a T, especially if Typescript already gives us [a way to detect both](https://medium.com/@teamcode20233/understanding-double-question-mark-operator-in-typescript-be869f210fe6).

I think, for the use case here, it makes sense for `signOut` to return `null` if there is no error. If we used `undefined`, it would sort of be saying that there was an error in getting the error, if that makes any sense at all.

It also makes sense for `session`, `userId`, and `userEmail` to be optionally `undefined`, because we haven't fetched that data yet. We can make an argument for `null` if the field can be fetched and is empty, but I don't think we'll encounter that error at least for `userId` and `userEmail`. Maybe we can look into `session` 🤔

Reply:
side note: we can reduce the definition to be `session?: Session`, which is Typescript's way of writing the same thing.

Reply:
yeah, that makes sense. It does seem like there are cases where `null` makes sense over `undefined`, and we should probably try and be consistent across the app as you mentioned

Reply:
Sounds good! I've replaced it with `session?: Session`